### PR TITLE
[oneDNN] Fix to issue #34554

### PR DESCRIPTION
**Pull request by jczaja**
### PR types
Bug fixes

### PR changes
OPs

### Describe
It is a fix to #34554 where models are using inplace execution. This inplace execution does not work well with caching mechanism so this PR is disabling caching for inplace capable ops: softmax, elementwise_*** and activation.

### CODE REVIEW

**Comment by jczaja** [comment]
@chenwhql Could you please approve PR-CI-APPROVAL. Its failure is a false positive due to fact that mechanism of checking PADDLE_ENFORCE_** works only on added lines of code. If we modify existing PADDLE_ENFORCE_** that spawn on multiple lines then CI mechanism will only see part of this call and pattern of VALID check may fail.

**Comment by lidanqing-vv** [comment]
@OliverLPH Could you please review it?

**Review by lidanqing-vv — DISMISSED** [review]
LGTM.

File: paddle/fluid/platform/mkldnn_reuse.h [hunk 1]
```diff
@@ -34,6 +34,211 @@ using framework::Tensor;
...
+    return this->AcquireMemoryFromPrimitive(bwd_pd_->diff_src_desc(), ptr);
+  }
+
+  // Buffer of given Tensor is used for oneDNN computation
+  std::shared_ptr<mkldnn::memory> AcquireDiffWeightsMemory(
+      framework::Tensor* diff_weights) {
+    PADDLE_ENFORCE_NOT_NULL(
+        bwd_w_pd_,
+        platform::errors::Unavailable(
+            "Error: BWD_W_PD should be set when getting BWD grad of weights."));
```
Comment (chenwhql) [inline paddle/fluid/platform/mkldnn_reuse.h]: `Error:` here is needless, other similar cases are same

File: paddle/fluid/platform/mkldnn_reuse.h [hunk 2]
```diff
@@ -34,6 +34,211 @@ using framework::Tensor;
...
+    return std::make_shared<TForward>(*fwd_pd_);
+  }
+
+  std::shared_ptr<TBackward> AcquireBackwardPrimitive() {
+    return std::make_shared<TBackward>(*bwd_pd_);
+  }
+
+  std::shared_ptr<TBackward_params> AcquireBackwardWeightsPrimitive() {
+    PADDLE_ENFORCE_NOT_NULL(bwd_w_pd_, platform::errors::Unavailable(
+                                           "Error: BWD_PD should be set when "
```
Comment (chenwhql) [inline paddle/fluid/platform/mkldnn_reuse.h]: I am not very sure.. if user see this error message, what changes they need to make? does the user understand BWD_PD?

File: paddle/fluid/operators/mkldnn/softmax_mkldnn_op.cc [hunk 3]
```diff
@@ -32,69 +32,54 @@ using platform::to_void_cast;
...
-                                              data_softmax_md, axis);
-      this->AcquireBackwardPrimitiveDescriptor(diff_softmax_md, data_softmax_md,
-                                               axis);
-    }
+      : platform::MKLDNNHandlerNoCachingT<T, mkldnn::softmax_forward,
+                                          mkldnn::softmax_backward>(
+            mkldnn_engine, cpu_place) {
+    PADDLE_ENFORCE_EQ(
+        out_grad->dims(), in_x_grad->dims(),
+        platform::errors::InvalidArgument("The shape of softmax_grad's input "
```
Comment (chenwhql) [inline paddle/fluid/operators/mkldnn/softmax_mkldnn_op.cc]: we recommend that the expected value and actual value are also reported in the error message

File: paddle/fluid/platform/mkldnn_reuse.h [hunk 4]
```diff
@@ -589,70 +794,64 @@ class MKLDNNHandler {
...
-                                       MKLDNNMemoryFormat::any);
-
-      auto attributes = CreateAttributes(algo, scale_x, scale_y, scale_z);
-      this->AcquireForwardPrimitiveDescriptor(attributes, algo, src0_md,
-                                              src1_md, dst_md);
+                      float scale_x, float scale_y, float scale_z)
+      : platform::MKLDNNHandlerNoCachingT<T, dnnl::binary>(engine, cpu_place) {
+    PADDLE_ENFORCE_EQ(
+        x->layout(), DataLayout::kMKLDNN,
+        platform::errors::InvalidArgument("Wrong layout set for X tensor."));
```
Comment (chenwhql) [inline paddle/fluid/platform/mkldnn_reuse.h]: recommend give expected value and actual value

File: paddle/fluid/platform/mkldnn_reuse.h (hunk 2 above)
Comment (jczaja) [reply paddle/fluid/platform/mkldnn_reuse.h]: User of this API is a developer of oneDNN PaddlePaddle operator.This error should be understandable as it implies than some other calls of this API should be called before this one when implementing operator.

File: paddle/fluid/platform/mkldnn_reuse.h (hunk 4 above)
Comment (jczaja) [reply paddle/fluid/platform/mkldnn_reuse.h]: ok

File: paddle/fluid/operators/mkldnn/softmax_mkldnn_op.cc (hunk 3 above)
Comment (jczaja) [reply paddle/fluid/operators/mkldnn/softmax_mkldnn_op.cc]: ok

File: paddle/fluid/platform/mkldnn_reuse.h (hunk 1 above)
Comment (jczaja) [reply paddle/fluid/platform/mkldnn_reuse.h]: ok

File: paddle/fluid/operators/elementwise/mkldnn/elementwise_mkldnn_op.h [hunk 5]
```diff
@@ -47,13 +47,19 @@ class EltwiseMKLDNNKernel : public framework::OpKernel<T> {
...
+                                             ctx.GetPlace(), x, y, z, scale_x,
+                                             scale_y, scale_o);
 
     const auto src_x_memory = handler.AcquireSrcMemory(x);
     const auto src_y_memory = handler.AcquireSecondSrcMemory(y);
-    const auto dst_memory = handler.AcquireDstMemory(z);
+    // For Inplace src and and dst are the same memory object
+    // (jczaja) UT mechanics is testing inplace for this op
+    // regardless shapes, which is wrong when X is to be broadcasted as output
+    // is of bigger shape that X.
```
Comment (sfraczek) [inline paddle/fluid/operators/elementwise/mkldnn/elementwise_mkldnn_op.h]: Could you clarify this comment, please?

Comment (jczaja) [reply paddle/fluid/operators/elementwise/mkldnn/elementwise_mkldnn_op.h]: sure. Comment was updated.

**Review by sfraczek — APPROVED** [review]
LGTM

**Comment by jczaja** [comment]
@chenwhql  Please continue your review

**Review by chenwhql — APPROVED** [review]
LGTM for PADDLE_ENFORCE

**Comment by jczaja** [comment]
@chenwhql Please approve PR-CI-APPROVAL

PR-CI-APPROVAL is failing as I modified first and third line of PADDLE_ENFORCE_EQ (bold black):
  **PADDLE_ENFORCE_EQ(ct.Analyze(9), true,**
                    platform::errors::InvalidArgument(
                        **"Invalid number of cached oneDNN objects"));**

And problem is that script evaluating this change is only recieving modified lines . So it will not recieve :
   ` platform::errors::InvalidArgument(`
And then script evaluating if PADDLE_ENFORCE is correct will fail as It cannot find missing line.

**Review by lidanqing-vv — APPROVED** [review]
LGTM

**Comment by OliverLPH** [comment]
hi @jczaja , I compiled this PR and test https://github.com/PaddlePaddle/Paddle/issues/34554, there are still some models will crash.

when `test_groups` equals 3, 4, 5, 8, 9 will meet Segmentation Fault
```shell
./build/model_test --single_thread=false --single_instance=false --test_groups="3"
```
e5147405c135bfb3215ec7941f00c768

could you please try re-test these models in your environment?

**Comment by jczaja** [comment]
@OliverLPH  Hi, I will test given commandline and fix if there is any crash

**Comment by chenwhql** [comment]
@jczaja this commit cause paddle compile failed, ci passed may be related by ccatch, so we revert this pr temporarily
11ead905866180d7219cdec6d3945936

**Comment by jczaja** [comment]
@chenwhql Could you please share a full log of this build that failed?

**Comment by lidanqing-vv** [comment]
@jczaja @chenwhql  HiHi, I have that, I will share with you @jczaja

**Pull request merged**